### DISCUSSION TITLE
Grid search through environment variables

### DISCUSSION THREAD

**Discussion by sAviOr287** [facebookresearch/hydra#2286] [Q&A]
Hi all,

I am rather new to hydra but think the framework is awesome.

I have a system where I can spawn many workers with each CPU and GPU specification. Each worker can then be spawned separately and each has an environment variable that can be accessed through $ENV_VARIABLE.

The way I do grid search is usually inside the python script:

```python
worker_parameter = int(os.environ["WORKER_ID"])
seed_vec = [9001]
lr_vec = [1e-2, 1e-3, 1e-4]

# Create a grid search vector using the itertools package
vector_grid = list(
    itertools.product(
        seed_vec,
        lr_vec
    )
)

parameters = vector_grid[int(worker_parameter)]
logging.info("Parameters:")
logging.info(parameters)
logging.info(args)
invoke_main(parameters, args)
```

This will then spawn for example 3 different workers where each worker will look at the different hyper parameter config. This is currently using parsers etc  which i know is against what Hydra wants to do.
Hence I was wondering if there is a way to do this grid search without creating many many config files :)

Thanks a lot in advance for your help. The framework and saving files is great !

**Comment by Jasha10** [comment]
Yes, this is possible using Hydra's sweep functionality! See the docs on [Multi-run](https://hydra.cc/docs/tutorials/basic/running_your_app/multi-run/).

**Reply by sAviOr287** [reply]
Hi Jasha,

Thanks a lot for the quick reply. Do you mind pointing me toward where I would have to insert the environment variable?
I see that I can add a list in the config file to be able to sweep but where would I be able to tell the machine that a specific worker is supposed to work on a specific config? 

I never used configs before so rather new to this atm ...

**Reply by Jasha10** [reply]
You can use the [`oc.env`](https://omegaconf.readthedocs.io/en/2.1_branch/custom_resolvers.html#oc-env) OmegaConf resolver to lazily access environment variables from the composed `DictConfig` object:
```yaml
# foo.yaml
worker_parameter_str: ${oc.env:WORKER_ID}  # this resolves to the string '123'
worker_parameter_int: ${oc.decode:${oc.env:WORKER_ID}}  # this resolves to the int 123
```

> I have a system where I can spawn many workers with each CPU and GPU specification. Each worker can then be spawned separately and each has an environment variable that can be accessed through $ENV_VARIABLE.

Could you say a bit more about how your system works?

Will you be invoking the python app once on your local machine and want it to spawn a job on each of the many workers? Or were you thinking of invoking the app on each of the machines separately? Hydra's sweeper/multirun accommodates the former use-case via remote launchers (such as e.g. the [submitit launcher](https://hydra.cc/docs/plugins/submitit_launcher/)).

**Reply by sAviOr287** [reply]
yeah the idea would be to spawn the job on each of the worker and invoking the app on each machine separately.
The thing with my setup is that I can spawn many workers that run the same code at once. But the environment variable changes for each worker so i can exploit that to do grid search.
I have tried to use the slumitit launcher but unfortunately it seems that my system is not slurm 

`RuntimeError: Could not detect "srun", are you indeed on a slurm cluster?`

**Reply by Jasha10** [reply]
I see.
Give the `oc.env` resolver (mentioned in my previous comment) a try :)

**Reply by sAviOr287** [reply]
sorry for the really stupid question but once i added that in the YAML file
how would that help me to iterate over the configs in the python script?
I would still have to create the itertools grid as before i assume?
In that case wouldnt the saving structure of hydra get messed up i.e. in the different runs would have the same config saved no
Thanks again for the quick reply

**Reply by Jasha10** [reply]
Yes, you're right. Hmm...
Would you mind sharing a more complete example of how you are spawning the jobs? How does Hydra fit into your system? I'm wondering if it will be possible for you to leverage Hydra's sweeping capability without having to write a custom launcher plugin that "wraps" the worker-spawning system.
